kernel/runtime/cxx/runtime/cxxsupport.cc

```diff
@@ -49,23 +49,66 @@
 #include "runtime/PageMap.hh"
 #include "runtime/KernelMemory.hh"
 #include "runtime/SimpleCapAlloc.hh"
+#include "runtime/ProcessorAllocator.hh"
 #include "runtime/tls.hh"
 #include "runtime/futex.hh"
 #include "runtime/umem.hh"
+#include "runtime/thread-extra.hh"
 
 extern mythos::InvocationBuf* msg_ptr asm("msg_ptr");
 extern mythos::Portal portal;
 extern mythos::CapMap myCS;
 extern mythos::PageMap myAS;
 extern mythos::KernelMemory kmem;
-extern mythos::SimpleCapAllocDel capAlloc;
+extern mythos::SimpleCapAlloc< mythos::init::APP_CAP_START
+  , mythos::init::SIZE-mythos::init::APP_CAP_START> capAlloc;
+extern mythos::ProcessorAllocator pa;
 
 #ifndef NUM_CPUS
 #define NUM_CPUS (2)
 #endif
 
 #define PS_PER_TSC (0x0000000000000181)
 
+struct PthreadCleanerSemaphore{
+  PthreadCleanerSemaphore()
+    : flag(FREE)
+  {
+    //MLOG_ERROR(mlog::app, "PthreadCleaner");
+  }
+
+  enum state{
+    UNKNOWN = 0,
+    FREE = 1,
+    EXITED = 2
+  };
+
+  void exit(){
+    //MLOG_DETAIL(mlog::app, "PthreadCleaner exit", DVARhex(this), DVARhex(pthread_self()));
+    auto ec = flag.exchange(EXITED);
+    if(ec != FREE){
+      ASSERT(ec!=UNKNOWN);
+      mythos::syscall_signal(ec);
+    }
+  }
+
+  void wait(pthread_t t){
+    auto pcs = reinterpret_cast<PthreadCleanerSemaphore* >(t - (pthread_self() - reinterpret_cast<uintptr_t>(this)));
+    //MLOG_DETAIL(mlog::app, "PthreadCleaner wait", DVARhex(pcs), DVARhex(this), DVARhex(pthread_self()), DVARhex(t));
+    while(pcs->flag.load() != EXITED){
+      mythos::CapPtr exp = FREE;
+      if(pcs->flag.compare_exchange_weak(exp, mythos_get_pthread_ec_self())){
+        //MLOG_DETAIL(mlog::app, "PthreadCleaner going to wait");
+        mythos_wait();
+      }
+    }
```

Comment:
This function is supposed to be called by only one thread, right? It seems like this does busy waiting if another thread comes first.

Reply:
it is some kind of simple futex. try to register as waiter and then going to sleep (in a loop). added comments